src/ctap/credential_id.rs

```diff
@@ -0,0 +1,268 @@
+    {
+        return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+    }
+    Ok(())
+}
+
+impl CredentialId {
+    /// Encrypts the given private key, relying party ID hash, and cred protect policy into a credential ID.
+    ///
+    /// Other information, such as a user name, are not stored. Since encrypted credential IDs are
+    /// stored server-side, this information is already available (unencrypted).
+    pub(super) fn encrypt_to_bytes(
+        env: &mut impl Env,
+        private_key: &PrivateKey,
+        rp_id_hash: &[u8; 32],
+        cred_protect_policy: Option<CredentialProtectionPolicy>,
+    ) -> Result<Vec<u8>, Ctap2StatusCode> {
+        let mut payload = Vec::new();
+        let cbor = cbor_map_options! {
+          CborKeyHandleField::PrivateKey => private_key,
+          CborKeyHandleField::RpIdHash=> rp_id_hash,
+          CborKeyHandleField::CredProtectPolicy => cred_protect_policy,
+        };
+        cbor_write(cbor, &mut payload)?;
+        add_padding(&mut payload)?;
+
+        let aes_enc_key =
+            crypto::aes256::EncryptionKey::new(&env.key_store().key_handle_encryption()?);
+        let encrypted_payload = aes256_cbc_encrypt(env.rng(), &aes_enc_key, &payload, true)?;
+        let mut credential_id = encrypted_payload;
+        credential_id.insert(0, CBOR_CREDENTIAL_ID_VERSION);
+
+        let id_hmac = hmac_256::<Sha256>(
+            &env.key_store().key_handle_authentication()?,
+            &credential_id[..],
+        );
+        credential_id.extend(&id_hmac);
+        Ok(credential_id)
+    }
+
+    /// Returns None if
+    /// - the format does not match any known versions, or
+    /// - the HMAC test fails.
+    ///
+    /// For v0 (legacy U2F) the credential ID consists of:
+    /// - 16 bytes: initialization vector for AES-256,
+    /// - 32 bytes: encrypted ECDSA private key for the credential,
+    /// - 32 bytes: encrypted relying party ID hashed with SHA256,
+    /// - 32 bytes: HMAC-SHA256 over everything else.
+    ///
+    /// For v1 (ECDSA (algorithm -7)) the credential ID consists of:
+    /// -  1 byte : version number
+    /// - 16 bytes: initialization vector for AES-256,
+    /// - 32 bytes: encrypted ECDSA private key for the credential,
+    /// - 32 bytes: encrypted relying party ID hashed with SHA256,
+    /// - 32 bytes: HMAC-SHA256 over everything else.
+    ///
+    /// For v2 (EdDSA over curve Ed25519 (algorithm -8, curve 6)) the credential ID consists of:
+    /// -  1 byte : version number
+    /// - 16 bytes: initialization vector for AES-256,
+    /// - 32 bytes: encrypted Ed25519 private key for the credential,
+    /// - 32 bytes: encrypted relying party ID hashed with SHA256,
+    /// - 32 bytes: HMAC-SHA256 over everything else.
+    ///
+    /// For v3 (CBOR key handle) the credential ID consists of:
+    /// -  1 byte : version number
+    /// - 16 bytes: initialization vector for AES-256,
+    /// - 64 bytes * N: encrypted CBOR key handle
```

Comment:
Why multiples of 64 byte? Block length is 32 bytes. Also I thought we wanted a fixed size of 241 byte for the key handle, so we can reject everything else?

Reply:
Done, changed to 192